### ISSUE TITLE
[Question] pooling and aggregation operations

### ISSUE THREAD

**Issue by xiexbing** [meta-pytorch/torchrec#2257]
in the forward pass, in the table wise sharding, when pooling is executed? is it after alltoall communication? and executed on trainer local? where can I see the exact code in torchrec code base?

in the backward pass, in the table wise sharding, when sorting and aggregation is executed? can you please point the lines in code base.

**Comment by JacoCheung** [comment]
> in the forward pass, in the table wise sharding, when pooling is executed?

The pooling is done in FBGEMM_GPU, not in torchrec itself. For example: https://pytorch.org/FBGEMM/fbgemm_gpu-python-api/table_batched_embedding_ops.html

**Comment by JacoCheung** [comment]
The backward key sortion and gradient reduction are also done inside fbgemm_gpu.

**Comment by xiexbing** [comment]
Can you point out for the classes in fbgemm_gpu that covers the backward
sorting and gradient aggregation?

**Comment by JacoCheung** [comment]
It's written in cpp, specifically, cuda source code.
- [The sort](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/src/split_embeddings_utils/transpose_embedding_input.cu#L311-L327)
- The aggregation is done inside CUDA kernels. (The source code is a `templatized` file and not readable, you can build fbgemm_gpu on your own and check the generated files. ) For [example](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/codegen/training/backward/embedding_backward_split_kernel_warp_template.cu#L218) 

[This template file](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/codegen/training/backward/embedding_backward_split_host_template.cpp) defines the autograd function that embraces forward and backward entry. A [generated file](https://gist.github.com/JacoCheung/dea1c7e964db65d4b9417119e535088b#file-gen_embedding_backward_split_adagrad-cpp-L433) example from the template file.

**Comment by xiexbing** [comment]
sorry, a followup question for the forward pass communication and pooling.
I profiled the forward pass with both nsight and torch profiler. I see the
all2all calls at python level, and the low level NCCL calls as SendReceive
for all2all, but I didn't see any calls map to the pooling. Does the
pooling actually happen within SendReceive? or it is after SendReceive and
within all2all? or it is after all2all, but just not appropriate profiled.

**Comment by xiexbing** [comment]
another question is: in the forward pass, before all2all, how to remove the
duplicated embedding indices per table, and then run all2all with input
(unique indices) and output (the corresponding embedding vectors). please
point out the code. thanks.

**Comment by JacoCheung** [comment]
> sorry, a followup question for the forward pass communication and pooling. I profiled the forward pass with both nsight and torch profiler. I see the all2all calls at python level, and the low level NCCL calls as SendReceive for all2all, but I didn't see any calls map to the pooling. Does the pooling actually happen within SendReceive? or it is after SendReceive and within all2all? or it is after all2all, but just not appropriate profiled.

The pooling is done before all2all. Maybe this figure is clearer
![image](https://github.com/user-attachments/assets/ed784d80-0b2f-4067-b68a-fa62a29cd891)

The first all2all is for lookup keys which is fed into fbgemm and fbgemm will do pooling for you. The second all2all performs all2all on pooled embedding. 
So the pooling happens actually in CUDA kernels. 
![image](https://github.com/user-attachments/assets/80c28ad6-7199-4cf6-aa27-bf30734b7cb9)

**Comment by xiexbing** [comment]
Thanks for the explanation. It helps a lot.
Let me dig deeper to some details:
1, can I assume the first all2all actually communicate about the KJT per
batch? so after this all2all, each GPU with a shard of an embedding table
actually will launch a fbgemm kernel to do the lookup and pooling. and then
use the 2nd all2all to send the pooled embedding vectors back to the KJT
batch owner. If this assumption is correct, for row wise parallelism,
all2all is still used in forward pass, if a sample has the indices across
multiple shards (means the embedding vectors stored on different GPUs), how
to do pooling?
2. if the above assumption is correct.  why not have only one all2all to
let each KJT batch owner have the individual embedding vectors (based on
their keys) and let them launch cuda kernel to do pooling locally.

**Comment by JacoCheung** [comment]
One correction: For **RW**, there is no the 2nd all2all, instead, it should be a ReduceScatter. (Table-wise should contain all2all)


> 1, can I assume the first all2all actually communicate about the KJT per
batch?

AFAIK, Yes. But be noted that a KJT all2all usually is composed of several tensor all2all.

> if a sample has the indices across
multiple shards (means the embedding vectors stored on different GPUs), how
to do pooling?

There should be a collective. For rw sharding, it could be a [reduce scatter](https://github.com/pytorch/torchrec/blob/v0.8.0/torchrec/distributed/sharding/rw_sharding.py#L388-L400). 

>  why not have only one all2all to let each KJT batch owner have the individual embedding vectors (based on
their keys) and let them launch cuda kernel to do pooling locally.

Initially, each rank has its dp input. The first all2all is used for sending keys to the corresponding sharding for lookup. I don't think we can skip it.  As I clarified, the 2nd should be a RS for RW. 
The local lookup result from fbgemm is a dense tensor **which includes all bags**, so it's ready for RS. If the fbgemm does not do pooling and return a jagged tensor, then we need a jagged tensor all2all + local reduce.

**Comment by JacoCheung** [comment]
>  before all2all, how to remove the duplicated embedding indices per table.

For `embeddingbag`, I think there's no dedpulication. But for `embedding`, there is a [deduplication in input_dist](https://github.com/pytorch/torchrec/blob/v0.8.0/torchrec/distributed/embedding.py#L989-L1015). You need to explicitly set [use_index_dedup=True](https://github.com/pytorch/torchrec/blob/main/torchrec/distributed/embedding.py#L1237) for `EmbeddingCollectionSharder` to enable unique.

**Comment by xiexbing** [comment]
thanks for the clarification! very helpful.

**Comment by xiexbing** [comment]
sorry, forget to ask some details on the collective communication operators. Just for confirmation: in the forward pass, for the communication about embedding indices, there are multiple all2all calls, each call for a jagged tensor (e.g. one call for values, one call for lengths, one call for lengths per key). In both forward and backward pass, for the communications for embedding vectors, there is only one call for all batches across all keys and tables. if the 2nd statement is correct, when the system scales up, the data size of of the call will increase n^2 (n is the number of ranks in the system). Why not cut the single call to multiple calls with each call communicate about 64MB (for best network bandwidth efficiency).
